core/pta/veraison_attestation/veraison_attestation.c

```diff
@@ -0,0 +1,158 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2024, Institute of Information Security (IISEC)
+ */
+
+#include <base64.h>
+#include <kernel/pseudo_ta.h>
+#include <mempool.h>
+#include <pta_veraison_attestation.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "cbor.h"
+#include "hash.h"
+
+#define PTA_NAME "veraison_attestation.pta"
+
+#define MAX_KEY_SIZE 4096
+#define MAX_NONCE_SIZE 64
+#define TEE_SHA256_HASH_SIZE 32
+
+#define EAT_PROFILE "http://arm.com/psa/2.0.0"
+#define CLIENT_ID 1
+#define LIFECYCLE 12288
+#define MEASURMENT_TYPE "PRoT"
+#define SIGNER_ID_LEN 32
+#define INSTANCE_ID_LEN 33
+
+/* clang-format off */
+#define SIGNER_ID {                                        \
+	0xac, 0xbb, 0x11, 0xc7, 0xe4, 0xda, 0x21, 0x72,    \
+	0x05, 0x52, 0x3c, 0xe4, 0xce, 0x1a, 0x24, 0x5a,    \
+	0xe1, 0xa2, 0x39, 0xae, 0x3c, 0x6b, 0xfd, 0x9e,    \
+	0x78, 0x71, 0xf7, 0xe5, 0xd8, 0xba, 0xe8, 0x6b     \
+}
+#define INSTANCE_ID {                                      \
+	0x01, 0xce, 0xeb, 0xae, 0x7b, 0x89, 0x27, 0xa3,    \
+	0x22, 0x7e, 0x53, 0x03, 0xcf, 0x5e, 0x0f, 0x1f,    \
+	0x7b, 0x34, 0xbb, 0x54, 0x2a, 0xd7, 0x25, 0x0a,    \
+	0xc0, 0x3f, 0xbc, 0xde, 0x36, 0xec, 0x2f, 0x15,    \
+	0x08                                               \
+}
+/* clang-format on */
+
+static TEE_Result cmd_get_cbor_evidence(uint32_t param_types,
+					TEE_Param params[TEE_NUM_PARAMS])
+{
+	const uint8_t *nonce = params[0].memref.buffer;
+	const size_t nonce_sz = params[0].memref.size;
+	uint8_t *output_buffer = params[1].memref.buffer;
+	size_t *output_buffer_len = &params[1].memref.size;
+	const uint8_t *psa_implementation_id = params[2].memref.buffer;
+	const size_t psa_implementation_id_len = params[2].memref.size;
+	TEE_Result status = TEE_SUCCESS;
+
+	const char eat_profile[] = EAT_PROFILE;
+	const int psa_client_id = CLIENT_ID;
+	const int psa_security_lifecycle = LIFECYCLE;
+	const char measurement_type[] = MEASURMENT_TYPE;
+	const uint8_t signer_id[SIGNER_ID_LEN] = { SIGNER_ID };
+	const uint8_t psa_instance_id[INSTANCE_ID_LEN] = { INSTANCE_ID };
+
+	uint8_t measurement_value[TEE_SHA256_HASH_SIZE] = { 0 };
+	size_t b64_measurement_value_len = TEE_SHA256_HASH_SIZE * 2;
+	char b64_measurement_value[TEE_SHA256_HASH_SIZE * 2] = { 0 };
+
+	UsefulBufC ubc_cbor_evidence = { NULL, 0 };
+	UsefulBufC ubc_cose_evidence = { NULL, 0 };
+
+	if (param_types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+					   TEE_PARAM_TYPE_MEMREF_OUTPUT,
+					   TEE_PARAM_TYPE_MEMREF_INPUT,
+					   TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!nonce || !nonce_sz)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!output_buffer && *output_buffer_len)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* Calculate measurement hash of memory */
+	status = get_hash_ta_memory(measurement_value);
```

Comment:
The evidence produced by the PTA is attesting the memory contents of the calling TA, but there is no way to establish trust in the PTA in the first place.

From a chain-of-trust (CoT) perspective, the arrangement should rather be:
1. PSA RoT attests secure world, including PTA;
2. PTA chains to the PSA RoT attestation evidence to add its value-add.

There are a couple of ways to handle the composition (which boil down to “direct” or “delegated” modes), but all cases require bootstraping the CoT using the platform RoT.

This missing link makes the PTA float on unclear trust assumptions about its own veracity, which weakens its overall stance.

Reply:
You're absolutely right. I understand that it's necessary to authenticate the integrity of the PTA itself from a lower layer, such as the bootloader. However, at this stage, I believe there isn't a method implemented to achieve that.

Reply:
True.

Brainstorming mode on: if we use the "delegated" model, we could assume a PSA platform attestation is fed into the PTA at boot time. Such a token could be easily mocked.